src/ripple/unl/tests/qalloc.h

```diff
@@ -0,0 +1,396 @@
+#include <memory>
+#include <new>
+#include <sstream>
+
+namespace ripple {
+namespace test {
+
+namespace detail {
+
+template <class = void>
+class qalloc_impl
+{
+private:
+    class block
+    {
+    private:
+        std::size_t count_ = 0;
+        std::size_t bytes_;
+        std::size_t remain_;
+        std::uint8_t* free_;
+
+    public:
+        block* next;
+
+        block (block const&) = delete;
+        block& operator= (block const&) = delete;
+
+        explicit
+        block (std::size_t bytes);
+
+        void*
+        allocate (std::size_t bytes, std::size_t align);
+        
+        bool
+        deallocate();
+
+        template <class U>
+        static
+        block*&
+        stash (U const* u);
+    };
+
+    block* used_ = nullptr;
+    block* free_ = nullptr;
+
+public:
+    enum
+    {
+        block_size = 256 * 1024
+    };
+
+    qalloc_impl() = default;
+    qalloc_impl (qalloc_impl const&) = delete;
+    qalloc_impl& operator= (qalloc_impl const&) = delete;
+
+    ~qalloc_impl();
+
+    void*
+    allocate (std::size_t bytes, std::size_t align);
+
+    void
+    deallocate (void* p);
+};
+
+} // detail
+
+template <class T>
+class qalloc_type
+{
+private:
+    template <class U>
+    friend class qalloc_type;
+
+    std::shared_ptr<
+        detail::qalloc_impl<>> impl_;
+
+public:
+    using value_type = T;
+    using pointer = T*;
+    using const_pointer = T const*;
+    using reference = typename
+        std::add_lvalue_reference<T>::type;
+    using const_reference = typename
+        std::add_lvalue_reference<T const>::type;
+
+    template <class U>
+    struct rebind
+    {
+        using other = qalloc_type<U>;
+    };
+
+    qalloc_type (qalloc_type const&) = default;
+    qalloc_type& operator= (qalloc_type const&) = default;
+#ifndef _MSC_VER
+    qalloc_type (qalloc_type&& other) = default;
+    qalloc_type& operator= (qalloc_type&&) = default;
+#endif
+
+    qalloc_type();
+
+    template <class U>
+    qalloc_type (qalloc_type<U> const& u);
+
+    T*
+    allocate (std::size_t n);
+
+    template <class U>
+    U*
+    alloc (std::size_t n);
+
+    void
+    deallocate (T* p, std::size_t n);
+
+    template <class U>
+    void
+    dealloc (U* p, std::size_t n);
+
+    void
+    destroy (T* t);
+
+    template <class U>
+    bool
+    operator== (qalloc_type<U> const& u);
+
+    template <class U>
+    bool
+    operator!= (qalloc_type<U> const& u);
+};
+
+/** Allocator optimized for delete in temporal order.
+
+    This allocator is optimized for the case where objects
+    are deleted in approximately the same order that they
+    were created.
+*/
+using qalloc = qalloc_type<int>;
+
+//------------------------------------------------------------------------------
+
+namespace detail {
+
+template <class _>
+qalloc_impl<_>::block::block (std::size_t bytes)
+    : bytes_ (bytes - sizeof(*this))
+    , remain_ (bytes_)
+    , free_ (reinterpret_cast<
+        std::uint8_t*>(this + 1))
+{
```

Comment:
Right, and that should be okay, `next` is set when we insert into the linked list